File: observatorio/conectores/python/observatorio_conector.py
"""
Conector del Observatorio para aplicaciones Flask.

Archivo unico, opcional y apagado por defecto. La aplicacion sigue funcionando igual con o sin el.
Implementa el protocolo descrito en conectores/PROTOCOLO.md (version 1.0).

Uso:
    from observatorio_conector import crear_blueprint
    bp = crear_blueprint(
        secreto=os.environ.get("OBS_CONECTOR_SECRETO", ""),
        activo=os.environ.get("OBS_CONECTOR_ACTIVO") == "1",
        buscar_usuario=lambda usuario: ...,          # -> objeto o None
        verificar_clave=lambda registro, clave: ..., # -> bool
        buscar_por_identidad=lambda id_externo, correo, usuario: ...,  # -> objeto o None
        iniciar_sesion=lambda registro: ...,         # crea la sesion local (por ejemplo login_user)
        perfil=lambda registro: {"id": ..., "usuario": ..., "correo": ..., "nombre": ..., "rol": ...},
        conjuntos={"usuarios": ("Usuarios registrados", lambda desde: [ {...}, ... ])},
        url_inicio="/",
    )
    app.register_blueprint(bp, url_prefix="/observatorio-conector")
    # Si la aplicacion usa Flask-WTF CSRFProtect, eximir el blueprint: csrf.exempt(bp)
"""

import base64
import hmac
import hashlib
import json
import threading
import time

from flask import Blueprint, jsonify, redirect, request

VERSION = "1.0"
TOLERANCIA_SEG = 300
VIGENCIA_VISTOS_SEG = 15 * 60


class _Vistos:
    """Memoria de valores de un solo uso (nonces y jti) con caducidad. Suficiente para un proceso.
    Con varios procesos gunicorn conviene sustituirla por Redis; el resto del conector no cambia."""

    def __init__(self):
        self._d = {}
        self._l = threading.Lock()

    def usar(self, clave):
        ahora = time.time()
        with self._l:
            for k in [k for k, t in self._d.items() if ahora - t > VIGENCIA_VISTOS_SEG]:
                del self._d[k]
            if clave in self._d:
                return False
            self._d[clave] = ahora
            return True


def _firmar(secreto, partes):
    return hmac.new(secreto.encode("utf-8"), "\n".join(partes).encode("utf-8"), hashlib.sha256).hexdigest()


def _b64url_dec(s):
    s = s + "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s.encode("ascii"))


def crear_blueprint(*, secreto, activo, buscar_usuario, verificar_clave, buscar_por_identidad,
                    iniciar_sesion, perfil, conjuntos=None, url_inicio="/", bitacora=None):
    conjuntos = conjuntos or {}
    nonces = _Vistos()
    jtis = _Vistos()
    bp = Blueprint("observatorio_conector", __name__)

    def anotar(evento, **detalle):
        if bitacora:
            try:
                bitacora(evento, detalle)
            except Exception:
                pass

    def error(estado, mensaje):
        return jsonify({"error": mensaje}), estado

    @bp.before_request
    def _apagado():
        if not activo or not secreto:
            return error(404, "Conector no disponible")

    @bp.route("", methods=["GET"], strict_slashes=False)
    def sso():
        if request.args.get("accion") != "sso":
            return error(404, "Conector no disponible")
        token = request.args.get("token", "")
        i = token.rfind(".")
        if i < 0:
            return error(400, "Token mal formado")
        datos, firma = token[:i], token[i + 1:]
        if not hmac.compare_digest(firma, _firmar(secreto, ["SSO", datos])):
            anotar("sso_firma_invalida", ip=request.remote_addr)
            return error(401, "Firma no válida")
        try:
            p = json.loads(_b64url_dec(datos).decode("utf-8"))
        except Exception:
            return error(400, "Token mal formado")
        if not p.get("exp") or p["exp"] < int(time.time()):
            return error(401, "Token vencido")
        if not p.get("jti") or not jtis.usar(p["jti"]):
            return error(401, "Token ya utilizado")
        registro = buscar_por_identidad(p.get("id_externo"), p.get("correo"), p.get("usuario_externo"))
        if not registro:
            anotar("sso_usuario_no_encontrado", correo=p.get("correo"))
            return error(404, "Usuario no encontrado en esta aplicación")
        respuesta = iniciar_sesion(registro)
        anotar("sso_ok", correo=p.get("correo"), ip=request.remote_addr)
        return respuesta if respuesta is not None else redirect(url_inicio, code=303)

    @bp.route("", methods=["POST"], strict_slashes=False)
    def firmado():
        ts = request.headers.get("X-Obs-Timestamp", "")
        nonce = request.headers.get("X-Obs-Nonce", "")
        firma = request.headers.get("X-Obs-Signature", "")
        if not ts or not nonce or not firma:
            return error(401, "Falta la firma")
        try:
            if abs(int(time.time()) - int(ts)) > TOLERANCIA_SEG:
                return error(401, "Marca de tiempo fuera de rango")
        except ValueError:
            return error(401, "Marca de tiempo no válida")
        cuerpo = request.get_data(as_text=True) or "{}"
        try:
            datos = json.loads(cuerpo)
        except ValueError:
            return error(400, "JSON no válido")
        accion = str(datos.get("accion", ""))
        if not hmac.compare_digest(firma, _firmar(secreto, ["POST", accion, ts, nonce, cuerpo])):
            anotar("firma_invalida", accion=accion, ip=request.remote_addr)
            return error(401, "Firma no válida")
        if not nonces.usar(nonce):
            return error(401, "Petición repetida")

        if accion == "estado":
            return jsonify({"ok": True, "version": VERSION, "conjuntos": list(conjuntos.keys()),
                            "hora": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())})

        if accion == "verificar":
            registro = buscar_usuario(str(datos.get("usuario", "")).strip())
            if not registro or not verificar_clave(registro, str(datos.get("clave", ""))):
                anotar("verificar_fallida", usuario=datos.get("usuario"), ip=request.remote_addr)
                return jsonify({"ok": False})
            p = perfil(registro)
            anotar("verificar_ok", usuario=p.get("usuario"), ip=request.remote_addr)
            return jsonify({"ok": True, "id": str(p.get("id")), "usuario": p.get("usuario"),
                            "correo": p.get("correo"), "nombre": p.get("nombre"), "rol": p.get("rol")})

        if accion == "exportar":
            nombre = str(datos.get("conjunto", ""))
            if nombre not in conjuntos:
                return error(404, "Conjunto no definido")
            descripcion, consultar = conjuntos[nombre]
            filas = consultar(datos.get("desde"))
            anotar("exportar", conjunto=nombre, filas=len(filas))
            return jsonify({"ok": True, "conjunto": nombre, "descripcion": descripcion, "filas": filas})

        return error(400, "Acción no reconocida")

    return bp

File: observatorio/conectores/referencia-node/observatorio-conector.js
// Conector del Observatorio para aplicaciones Node.js. Archivo unico, opcional y apagado por defecto.
// La aplicacion sigue funcionando igual con o sin este archivo. Para retirarlo basta con dejar de montarlo.
//
// Uso (Express o http nativo):
//   import { crearConector } from './observatorio-conector.js';
//   const conector = crearConector({
//     activo: process.env.OBS_CONECTOR_ACTIVO === '1',
//     secreto: process.env.OBS_CONECTOR_SECRETO,
//     buscarUsuario: async (usuario) => ({ id, usuario, correo, nombre, rol, clave_hash }) | null,
//     verificarClave: async (registro, clave) => true | false,
//     buscarPorIdentidad: async ({ id_externo, correo }) => registro | null,
//     iniciarSesion: async (req, res, registro) => { /* crea la sesion local como lo hace el login normal */ },
//     urlInicio: '/',
//     conjuntos: { usuarios: { descripcion: 'Usuarios registrados', consultar: async () => [ {...} ] } },
//   });
//   app.all('/observatorio-conector', conector);       // Express
//   // o en http nativo: if (url.pathname === '/observatorio-conector') return conector(req, res);
import crypto from 'node:crypto';

const VERSION = '1.0';
const TOLERANCIA_SEG = 300;

export function crearConector(op) {
  const noncesVistos = new Map();
  const jtiVistos = new Map();
  const limpiar = (m) => { const ahora = Date.now(); for (const [k, t] of m) if (ahora - t > 15 * 60 * 1000) m.delete(k); };

  const firmar = (partes) => crypto.createHmac('sha256', op.secreto).update(partes.join('\n')).digest('hex');
  const iguales = (a, b) => a && b && a.length === b.length && crypto.timingSafeEqual(Buffer.from(a), Buffer.from(b));
  const json = (res, estado, obj) => { res.writeHead(estado, { 'Content-Type': 'application/json; charset=utf-8', 'Cache-Control': 'no-store' }); res.end(JSON.stringify(obj)); };

  function leerCuerpo(req) {
    if (req.body && typeof req.body === 'object') return Promise.resolve(JSON.stringify(req.body));
    return new Promise((ok, fallo) => { let d = ''; req.on('data', (c) => { d += c; if (d.length > 1e6) req.destroy(); }); req.on('end', () => ok(d)); req.on('error', fallo); });
  }

  return async function conector(req, res) {
    if (!op.activo || !op.secreto) return json(res, 404, { error: 'Conector no disponible' });
    const url = new URL(req.url, 'http://local');

    // --- Apertura de sesion desde el portal (GET con token de un solo uso) ---
    if (req.method === 'GET' && url.searchParams.get('accion') === 'sso') {
      const token = url.searchParams.get('token') || '';
      const i = token.lastIndexOf('.');
      if (i < 0) return json(res, 400, { error: 'Token mal formado' });
      const datos = token.slice(0, i), firma = token.slice(i + 1);
      if (!iguales(firma, firmar(['SSO', datos]))) return json(res, 401, { error: 'Firma no válida' });
      let p; try { p = JSON.parse(Buffer.from(datos, 'base64url').toString('utf8')); } catch { return json(res, 400, { error: 'Token mal formado' }); }
      if (!p.exp || p.exp < Math.floor(Date.now() / 1000)) return json(res, 401, { error: 'Token vencido' });
      limpiar(jtiVistos); if (jtiVistos.has(p.jti)) return json(res, 401, { error: 'Token ya utilizado' }); jtiVistos.set(p.jti, Date.now());
      const registro = await op.buscarPorIdentidad({ id_externo: p.id_externo, correo: p.correo, usuario: p.usuario_externo });
      if (!registro) return json(res, 404, { error: 'Usuario no encontrado en esta aplicación' });
      await op.iniciarSesion(req, res, registro);
      if (!res.headersSent) { res.writeHead(303, { Location: op.urlInicio || '/' }); res.end(); }
      return;
    }

    // --- Llamadas firmadas del portal (POST JSON) ---
    if (req.method !== 'POST') return json(res, 405, { error: 'Método no permitido' });
    const ts = req.headers['x-obs-timestamp'], nonce = req.headers['x-obs-nonce'], firma = req.headers['x-obs-signature'];
    if (!ts || !nonce || !firma) return json(res, 401, { error: 'Falta la firma' });
    if (Math.abs(Math.floor(Date.now() / 1000) - Number(ts)) > TOLERANCIA_SEG) return json(res, 401, { error: 'Marca de tiempo fuera de rango' });
    limpiar(noncesVistos); if (noncesVistos.has(nonce)) return json(res, 401, { error: 'Petición repetida' });
    const cuerpo = await leerCuerpo(req);
    let datos; try { datos = JSON.parse(cuerpo || '{}'); } catch { return json(res, 400, { error: 'JSON no válido' }); }
    const accion = datos.accion;
    if (!iguales(firma, firmar(['POST', accion, ts, nonce, cuerpo]))) return json(res, 401, { error: 'Firma no válida' });
    noncesVistos.set(nonce, Date.now());

    if (accion === 'estado') return json(res, 200, { ok: true, version: VERSION, conjuntos: Object.keys(op.conjuntos || {}), hora: new Date().toISOString() });

    if (accion === 'verificar') {
      const registro = await op.buscarUsuario(String(datos.usuario || ''));
      if (!registro) return json(res, 200, { ok: false });
      const bien = await op.verificarClave(registro, String(datos.clave || ''));
      if (!bien) return json(res, 200, { ok: false });
      return json(res, 200, { ok: true, id: String(registro.id), usuario: registro.usuario, correo: registro.correo, nombre: registro.nombre, rol: registro.rol });
    }

    if (accion === 'exportar') {
      const c = (op.conjuntos || {})[datos.conjunto];
      if (!c) return json(res, 404, { error: 'Conjunto no definido' });
      const filas = await c.consultar({ desde: datos.desde || null });
      return json(res, 200, { ok: true, conjunto: datos.conjunto, descripcion: c.descripcion, filas });
    }

    return json(res, 400, { error: 'Acción no reconocida' });
  };
}
